### ISSUE TITLE
glfwPostEmptyEvent sometimes fails to unblock glfwWaitEvents on Ubuntu 18.04

### ISSUE THREAD

**Issue by OlivierSohn** [glfw/glfw#1281]
From times to times, the empty event posted by a call to `glfwPostEmptyEvent` won't unblock a call to `glfwWaitEvent` in the main thread.

A minimal program reproducing the issue [is available here](https://github.com/OlivierSohn/repro-evt)
It can be built and run on linux like so:
```g++ main.cpp -std=c++14 -lglfw -lX11 -ldl -lpthread && ./a.out```

If the calls of `glfwPostEmptyEvent` and `glfwWaitEvent` don't overlap in time, then the bug will never happen, but if the 2 calls are concurrent, the bug happens, "sometimes".

Note that on OSX, this bug never happens.

GLFW Version (from glfwGetVersionString()):
`3.2.1 X11 GLX EGL clock_gettime /dev/js Xf86vm shared`

Ubuntu version
`18.04 LTS`

- Note that I have tried posting "several" empty events at the same time: it doesn't help either, it just makes the bug a little less reproducible but doesnt fix it entirely.

Edit : added the build command
Edit2 : fixed the build command

**Comment by OlivierSohn** [comment]
The bug also exists with the latest version of the `master` branch

**Comment by OlivierSohn** [comment]
I also tried to fix the bug by changing the implementations of `glfwPostEmptyEvent` and `glfwWaitEvents` for X11, but could only make the bug "lees frequent", so it's not a solution.

My workaround sofar is to systematically post a _second_ empty event, one millisecond after the first one, to have the guarantee that the freeze will not exceed a millisecond when it happens.

**Comment by kovidgoyal** [comment]
Wouldn't the fix be simply to use a mutex either wrapping your calls to glfwPostEmptyEvent() and glfwWaitEvents() or add a mutex to glfw itself (which may not be appropriate since glfw does not depend on ptthreads, IIRC)

**Comment by OlivierSohn** [comment]
@kovidgoyal if I wrap my calls to glfwPostEmptyEvent() and glfwWaitEvents() in a mutex, I won't be able, ever, to wake up the thread that made the glfwWaitEvents() call by calling glfwPostEmptyEvent() from another thread, because ... of the mutex :) so it's not a solution.

Using a mutex at a finer level inside glfw is an approach that could work, my first attempt was not conclusive, but there might be a possible solution there. Do you want to give it a try?

**Comment by kovidgoyal** [comment]
Oh yeah, sorry, I wasn't paying much attention to the context :) Thinking about it some more, I dont think this will work, or at least, one needs to understand the root cause of the failure first. What is happening to the events sent by XSendEvent? Why aren't they being picked up by XNextEvent?

**Comment by OlivierSohn** [comment]
> with the i3 window manager, it freezes before starting the test

Can you add logs to the test program to see where it freezes?

If I run your program on my system, it works fine, so that shows that we need a window to reproduce the problem.

**Comment by kovidgoyal** [comment]
It freezes on the window creation step, basically glfwCreateWindow() never returns

If you need a window to reproduce it, it means something on your system is interfering with event handling, probably accessing the X event queue itself. Some kind of plugin that hooks into all processes? Ubuntu is notorious for those.

**Comment by OlivierSohn** [comment]
> It freezes on the window creation step, basically glfwCreateWindow() never returns

I think I'm using "the standard way" of creating a glfw window, so I'm surprised that it doesn't work for you.

> I can reproduce with your example, by adding glfwWindowHint(GLFW_VISIBLE, false);
before the call to the window creation step.

I don't see why this call is necessary, it seems to me it's an unrelated issue (an issue nonetheless).

> If you need a window to reproduce it, it means something on your system is interfering with event handling, probably accessing the X event queue itself. Some kind of plugin that hooks into all processes? Ubuntu is notorious for those.

The window itself probably needs the X event queue, for mouse events, resize events, etc... so I'm not surprised that having or not having a window can make a difference.

**Comment by kovidgoyal** [comment]
The reason the window is needed is because of line 1115 in window.c Basically glfwpostemptyevent is a no-op when no windows are present. I dont know why that is the case, probably needed by some other backend.

**Comment by OlivierSohn** [comment]
> one needs to understand the root cause of the failure first. What is happening to the events sent by XSendEvent? Why aren't they being picked up by XNextEvent?

Let me explain what I understood about "what can be the root cause" :

I think the problem is that to wait for events, we use `select` (see the comment on top of waitForEvent in x11_window.c) instead of using X-related functions.

X-related functions use the display-lock when necessary to avoid race conditions, but select doesn't.

**Comment by kovidgoyal** [comment]
That should be easy to check, just replace the implementation of waitForEvent with XEventsQueued see if that fixes it.

**Comment by kovidgoyal** [comment]
Using XPeekEvent instead of the select does indeed fix the issue on my system

**Comment by kovidgoyal** [comment]
And this commit is the one that introduced the use of select, as best as I can tell: https://github.com/glfw/glfw/commit/eb7688df8f851b2a21047f757096b500ded34e48

I dont see any good way to fix this. As far as I can tell, you cannot both be lock free and avoid race conditions. I dont know why it is important to be able to use GLX calls from another thread, OpenGL calls, sure, but GLX?

**Comment by kovidgoyal** [comment]
CPU consumption would be awful with that. The other possibility is to use a dedicated extra fd pair. Have the select call wait on that as well and have the postemptyevent function write to it.

**Comment by OlivierSohn** [comment]
so `XPending()` + `sched_yield()` works (the race condition is solved), we are lock-free (i.e one can take the display lock when we release it), but the problem is we use 100%cpu! we could use usleep(...) instead of sched_yield() but that would increase latency...

> I dont know why it is important to be able to use GLX calls from another thread, OpenGL calls, sure, but GLX?

I have no idea

**Comment by kovidgoyal** [comment]
See my previous post.

**Comment by OlivierSohn** [comment]
> The other possibility is to use a dedicated extra fd pair. Have the select call wait on that as well and have the postemptyevent function write to it.

Sounds good

**Comment by kovidgoyal** [comment]
I'm happy to implement that in my glfw fork (and switch to using poll as well instead of select) since this issue could affect me, though I have not seen it in the wild so far. However you'll need to get buy in from @elmindreda if you want to get the fix merged here.

**Comment by OlivierSohn** [comment]
Thanks, it would be great! If you want I can cherrypick your commit to propose a merge here afterwards (or did you intend to merge your fork?).

Just to give some context, in my games I use event-driven loops : no rendering happens until an event (coming from the server, or from the platform) occurs. I'm not sure I would have noticed this bug had I used a time-based loop where every 10 milli seconds a render occurs...

**Comment by kovidgoyal** [comment]
I use an event based loop as well, isn't that what everybody uses? Why would you render repeatedly if nothing has happened? As for merging my fork, it seems unlikely at this point. It has pretty much diverged since #1140 and tons f other bug fixes/enhancements.

Feel free to cherry pick and propose a merge, I have no objections.

**Comment by OlivierSohn** [comment]
I should have added this to the context description : in my case, I also have different sources of events : glfw events (which I listen to on the main thread), and server events (which are listened to on another thread). If the server event arrives before the glfw event, then to continue on the main thread I need to wakeup the glfw wait function (glfwWaitEvents) using glfwPostEmptyEvent.

Hence, this bug becomes very visible :)

**Comment by kovidgoyal** [comment]
Yeah same architecture for me (in my case there are two secondary threads listening for server events and events from child processes), but I've never run into this in the wild and kitty (my project that uses glfw) has been out in the wild for almost two years now.

But anyway, I agree with your point, this is a potentially bad bug and should be fixed.

**Comment by OlivierSohn** [comment]
IMO, https://github.com/kovidgoyal/kitty/issues/426 may be related to this. Quoting the last comment:

> They stay that way, until I do something to snap them out of it. Switching desktops or sometimes even just pressing shift (!) seems to do it, or just moving the mouse

This is the same behaviour I saw in my games : if I moved the mouse or pressed a key, then that would wake up the `glfwWaitEvents` call.

But I might be wrong.

**Comment by kovidgoyal** [comment]
Seems unlikely this would cause high cpu usage, though. But, with threading bugs, who knows :)

**Comment by OlivierSohn** [comment]
I was wondering that too, maybe it's just that one empty event post is missing somewhere in the code (I guess it's hard to debug without having a repro case...)

**Comment by kovidgoyal** [comment]
I have committed a fix. Your test case passes and kitty seems to be unaffected by it, fee, free to review.

**Comment by OlivierSohn** [comment]
Cool! I'm looking at it

**Comment by OlivierSohn** [comment]
I made a comment on the commit, and also, what about using `ppoll` instead of `poll` which has a better time precision ?

**Comment by kovidgoyal** [comment]
I dont know how portable ppoll is

**Comment by OlivierSohn** [comment]
I was just wondering if that would be seen as a regression wrt `select` time resolution. I think to be sure we need to benchmark, to see what the implications are... but as a first approximation it should work. I made another comment on the new commit, to simplify the code, you might want to check it out.

**Comment by kovidgoyal** [comment]
I;ve never come across a need for sub-millisecond timeouts. For example, QTimer from Qt uses millisecond resolution timers. And glfw itself makes no guarantees about timer resolution. The wayland backend uses milliseconds as well.

**Comment by OlivierSohn** [comment]
Another thought, what happens if we repeateadly call glfwPostEmptyEvent, without calling glfwPollEvents or glfwWait***Events : at some point the self pipe will overflow? block? crash?

**Comment by kovidgoyal** [comment]
Its non-blocking, so the writes will simply fail. The write failure is ignored, so it is harmless.

**Comment by OlivierSohn** [comment]
@kovidgoyal yes, retrying the write could be nice, if someone uses signals

**Comment by kovidgoyal** [comment]
Done. And just for completeness, here is my analysis of the root cause:

basically, depending on the exact timing of the calls, it can happen that Xlib on the GUI thread reads data of the fd and stores it in its internal cache, which causes the select() to block, even though the other thread has sent an event that the main thread does not yet see.

This is fundamentally because of the bad (or perhaps more correct ancient) design of Xlib. it really is is not made for multiplexed I/O

**Comment by OlivierSohn** [comment]
I was going to open a PR, and did some benchmarks first:

on Ubuntu / X11 (this is the scope of the fix):

- current implementation where we use `select` to wait with timeout:
  - glfwWaitEventsTimeout has a precision of 80 micro-seconds
- with the fix, where we use `poll` to wait with timeout:
  - glfwWaitEventsTimeout has a precision of 1100 micro-seconds

To me, as a user of the library, this change sounds like bad news. Of course, the games I develop will continue to run with the decreased precision, but the timing we be a _little_ more "off".

@kovidgoyal do you think it's possible to do the same idea but with select? what are the disadvantages of select that made you chose poll? I don't mind doing the implementation with select.

Also, for completeness, I will benchmark with `ppoll`

**Comment by kovidgoyal** [comment]
Just google select vs poll. Basically select becomes highly inefficient if the fd number of one of the monitored fds happens to be high.

**Comment by OlivierSohn** [comment]
So the benchmark of `glfwWaitEventsTimeout` with `ppoll` is similar to what happens for `select` : 80micro second precision, except that below that, it stays at 80, instead of dropping to 0 for `select`.

The problem is that it seems less portable than other functions so we can't really use that...

**Comment by kovidgoyal** [comment]
And note that the wayland backend use poll() anyway. Which reminds me that this fix probably needs to be ported to the wayland backend as well.

**Comment by OlivierSohn** [comment]
And on OSX, the precision of `glfwWaitEventsTimeout` is also in the 80 micros. So I didn't test with wayland but it seems it would currently be 1 millisecond then? Maybe we can see that as a bug and fix it by using select instead?

**Comment by OlivierSohn** [comment]
Note that to do the benchmarks I use https://github.com/OlivierSohn/repro-evt/blob/master/timeout.cpp

**Comment by kovidgoyal** [comment]
As far as i can see, ppoll is available on linux and freebsd and openbsd and on netbsd under the name pollts, so it's probably fine to use it. But as I dont need sub-millisecond timers, I'm not particularly motivated to make that change.

**Comment by OlivierSohn** [comment]
I did the change, it's working, the only thing is I have a compilation warning because the function is not found in the header <poll.h>, i tried to fix it by passing -D_GNU_SOURCE but it failed. I'll look into it a bit more then, thanks for doing the research on portability!

**Comment by kovidgoyal** [comment]
Never mind, I made the change.

**Comment by OlivierSohn** [comment]
Btw, regarding the availability of ppoll it would be nice if we can document from which version of the different OSes it is available? I'll try to do my research on that...

Also, did you intentionally leave the non-timeout call using `poll`? I was thinking that we can use ppoll here too, to be consistent with the timeout case (passing NULL instead of -1 for the time indicates "waiting forever").

**Comment by kovidgoyal** [comment]
Feel free to add documentation for version availability. There is no need to use ppoll when timeout is not specified, so I didn't change it.

**Comment by kovidgoyal** [comment]
Your test case was failing on wayland as well, so I ported the fix to the wayland backend and cleaned it up a little removing the externs and using pipe2, since pipe2 is available everywhere that ppoll is available.

Please review.

**Comment by OlivierSohn** [comment]
See my previous comment :)

**Comment by OlivierSohn** [comment]
I also made a new comment on the last commit.

**Comment by OlivierSohn** [comment]
@kovidgoyal  It's hard to follow comments on commits, so I'll write it here (ideally we should discuss this on a PR):

- I think there is some code that could be factored between the x11 and wayland backends, especially in the new code. One way to share the code would be to write functions in a source file, that we `#include` in both backends. This way, when we make a change / review other people's code, we don't have to double check everything, which is error-prone. For example the deinitialization code now has 2 different behaviours and there is no way to know if it's because something was forgotten to be changed or because it was intended to be this way. (see next point)
- in one of the 2 backends, the pipes fd are closed only if the display is not null, is it intentional? If so, please add a comment saying why.

**Comment by OlivierSohn** [comment]
I'll take a shot at refactoring the code to use more functions to encapsulate the code, and to share code between backends.

**Comment by elmindreda** [comment]
That's a lot of detective work, thank you! :purple_heart:

I'm currently finishing Windows 10 per-monitor v2 DPI support. Will have a proper look at this as soon as I'm able. PR noted.

The GLX call that's expected on non-main threads is `glXSwapBuffers`.

**Comment by OlivierSohn** [comment]
@elmindreda you're welcome, I wanted to get to the bottom of this bug to see what was going on exactly, eventhough I had a working workaround...

**Comment by kovidgoyal** [comment]
Just FYI, I greatly cleaned up the implementation of this in my fork. The X11/Wayland event loops now support watching an arbitrary number of fds as well as an arbitrary number of timers. This was needed to integrate DBUS into the event loops which in turn was needed for IME support #41

**Comment by OlivierSohn** [comment]
@kovidgoyal could you please point out the relevant commits in your fork? I'd like to have an idea of what the incremental changes are, if we want to report the changes here. Thanks!

**Comment by kovidgoyal** [comment]
There is only one commit in my fork, the latest https://github.com/kovidgoyal/glfw/commit/f4b72f2492d250816775a7ce8dc2ab89637eacb4

If you want to see the code under development look at the commits under the glfw sub-directory of kitty.

**Comment by OlivierSohn** [comment]
Thanks. imo it's a change that should not be part of this particular issue. But it might be relevant for other issues or features.

**Comment by annacrombie** [comment]
I wonder if I'm running into this issue.  I'm running glfw on wayland, and once in awhile my program randomly freezes.  Attaching with gdb shows this backtrace:

```
#0  __cp_end () at src/thread/x86_64/syscall_cp.s:29
#1  0x00007fafcbcacc6d in __syscall_cp_c (nr=0, u=8, v=140721514594888, w=<optimized out>,
    x=<optimized out>, y=<optimized out>, z=0) at src/thread/pthread_cancel.c:33
#2  0x00007fafcbcb2b16 in read (fd=<optimized out>, buf=<optimized out>, count=<optimized out>)
    at src/unistd/read.c:6
#3  0x000055777d5240c8 in handleEvents ()
#4  0x000055777d524c8e in _glfwPlatformPollEvents ()
#5  0x000055777d51d825 in glfwPollEvents ()
#6  0x000055777d4fbb18 in opengl_ui_handle_input (ctx=0x7fafcbc4bde0, km=0x7ffc47e3c388,
    hf=0x7fafcbc4bc80) at ../client/ui/opengl/ui.c:125
#7  0x000055777d4ee559 in ui_handle_input (ctx=0x7fafcbc4bda0, km=0x7ffc47e3c388,
    hf=0x7fafcbc4bc80) at ../client/ui/common.c:82
#8  0x000055777d4eda24 in main (argc=3, argv=0x7ffc47e3eb78) at ../client/main.c:65
```
